src/storage/v2/disk/storage.cpp

```diff
@@ -868,7 +869,40 @@ DiskStorage::DiskAccessor::DetachDeleteVertex(VertexAccessor *vertex) {
       std::move(deleted_edges));
 }
 
-void DiskStorage::DiskAccessor::PrefetchEdges(const auto &prefetch_edge_filter) {
+bool DiskStorage::DiskAccessor::PrefetchEdgeFilter(const std::string_view &diskEdgeKey,
+                                                   const VertexAccessor &vertex_acc, bool isOutEdge) {
+  // diskEdgeKey format: vertex_out_gid | vertex_in_gid | direction | edge_type | GID
+  std::string_view disk_vertex_gid;
+  auto edges_res = (isOutEdge ? vertex_acc.OutEdges(storage::View::NEW) : vertex_acc.InEdges(storage::View::NEW));
+  if (isOutEdge) {
+    disk_vertex_gid = diskEdgeKey.substr(0, diskEdgeKey.find('|'));
+  } else {
+    auto vertex_in_start = diskEdgeKey.find('|') + 1;
+    disk_vertex_gid = diskEdgeKey.substr(vertex_in_start, diskEdgeKey.find('|', vertex_in_start) - vertex_in_start);
+  }
+  auto edge_gid = diskEdgeKey.substr(diskEdgeKey.rfind('|') + 1);
+
+  if (edges_res.HasValue()) {
+    for (const auto &edge_acc : edges_res.GetValue()) {
+      if (utils::SerializeIdType(edge_acc.Gid()) == edge_gid) {
+        // We already inserted this edge into the vertex's in_edges list.
+        return false;
+      }
+    }
+  }
+
+  // We need to iterate through edges_to_delete_ also because removed edges are not presented in edges_res
+  for (auto &deleted_edge_key : edges_to_delete_) {
+    if (deleted_edge_key == diskEdgeKey) {
+      // We already deleted this edge.
+      return false;
+    }
+  }
+
+  return disk_vertex_gid == utils::SerializeIdType(vertex_acc.Gid());
```

Comment:
but we can do this check as soon as we assert that edges exist. Because it doesn't make sense to do 2 for loops and then at the do the check with disk_vertex_gid.